# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [1]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()
from newsapi.newsapi_client import NewsApiClient

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\Kevin\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!

Bad key savefig.frameon in file D:\anaconda3\envs\pyvizenv\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle, line 421 ('savefig.frameon : True')
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.3.4/matplotlibrc.template
or from the matplotlib source distribution

Bad key verbose.level in file D:\anaconda3\envs\pyvizenv\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle, line 472 ('verbose.level  : silent      # one of silent, helpful, debug, debug-annoying')
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.3.4/matplotlibrc.template
or from the matplotlib source distribution

Bad key verbose.fileo in file D:\anaconda3\envs\pyvizenv\lib\site-packages\matplotlib\mpl-d

In [2]:
# Read your api key environment variable
load_dotenv()
api_key = os.getenv("NEWS_API_KEY")

In [3]:
# Create a newsapi client
newsapi = NewsApiClient(api_key = api_key)

In [4]:
# Fetch the Bitcoin news articles
btc_news = newsapi.get_everything(q = "bitcoin AND 2021", language = "en")

In [5]:
# Fetch the Ethereum news articles
eth_news = newsapi.get_everything(q = "ethereum AND 2021", language = "en")

In [51]:
# Create the Bitcoin sentiment scores DataFrame
btc_sentiments = []

for article in btc_news["articles"]:
    try:
        title = article["title"]
        description = article["description"]
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        positive = sentiment["pos"]
        negative = sentiment["neg"]
        neutral = sentiment["neu"]

        btc_sentiments.append({"text": text, "date": date, "compound": compound, "positive": positive, "negative": negative, "neutral": neutral})
    except AttributeError:
        pass

btc_sent_df = pd.DataFrame(btc_sentiments)
btc_df = btc_df.merge(btc_sent_df)

cols = ["date", "title", "description", "text", "compound", "positive", "neutral", "negative"]
btc_df = btc_df[cols]

btc_df.head()

date                                              title  \
0  2021-05-07  What Square’s smashing earnings tell us about ...   
1  2021-05-12  Tesla stops taking Bitcoin for vehicle purchas...   
2  2021-04-13  African crypto usage spurs Luno as customers r...   
3  2021-05-11  Mark Zuckerberg has a goat named Bitcoin and w...   
4  2021-05-12  Elon Musk says Bitcoin no longer accepted for ...   

                                         description  \
0  Today we're talking Square earnings and its bi...   
1  Tesla has stopped accepting bitcoin as payment...   
2  The crypto industry as a whole has seen a mome...   
3  Mark Zuckerberg posted a picture of his two, f...   
4  Tesla CEO Elon Musk has changed his stance on ...   

                                                text  compound  positive  \
0  Shares of Square are up more than 6% today aft...    0.2960     0.058   
1  Image: Tesla\r\n\n \n\n Tesla has stopped acce...    0.4939     0.134   
2  The crypto industry as a whole has seen a mome...    0.6124     0.135   
3  Mark Zuckerberg posted a picture of his two, f...    0.8455     0.217   
4  Tesla CEO Elon Musk has changed his stance on ...    0.3818     0.071   

   neutral  negative  
0    0.942      0.00  
1    0.816      0.05  
2    0.865      0.00  
3    0.783      0.00  
4    0.929      0.00

In [54]:
# Create the Bitcoin sentiment scores DataFrame
eth_sentiments = []

for article in eth_news["articles"]:
    try:
        title = article["title"]
        description = article["description"]
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        positive = sentiment["pos"]
        negative = sentiment["neg"]
        neutral = sentiment["neu"]

        eth_sentiments.append({"text": text, "date": date, "compound": compound, "positive": positive, "negative": negative, "neutral": neutral})
    except AttributeError:
        pass

eth_sent_df = pd.DataFrame(eth_sentiments)
eth_df = eth_df.merge(eth_sent_df)

cols = ["date", "title", "description", "text", "compound", "positive", "neutral", "negative"]
eth_df = eth_df[cols]

eth_df.head()

date                                              title  \
0  2021-05-04  Radical Ethereum entrepreneurs are redefining ...   
1  2021-04-20  Venmo now lets users buy and sell Bitcoin and ...   
2  2021-05-01  Move over, Bitcoin. Ethereum is at an all-time...   
3  2021-05-03              Ethereum breaks past $3,000 - Reuters   
4  2021-05-03               Ethereum breaks past $3000 - Reuters   

                                         description  \
0  Their investors call them disruptive innovator...   
1  Venmo is jumping aboard the cryptocurrency ban...   
2  Bitcoin prices continued their rebound Saturda...   
3  Cryptocurrency ether hit a record high of $3,0...   
4  Cryptocurrency ether hit a record high of $3,0...   

                                                text  compound  positive  \
0  Their investors call them disruptive innovator...   -0.2144     0.054   
1  Venmo is jumping aboard the cryptocurrency ban...    0.0258     0.034   
2  New York (CNN Business)Bitcoin prices continue...    0.0000     0.000   
3  By Reuters Staff\r\nFILE PHOTO: A representati...    0.0000     0.000   
4  A representation of virtual currency Ethereum ...    0.0000     0.000   

   neutral  negative  
0    0.835     0.111  
1    0.966     0.000  
2    1.000     0.000  
3    1.000     0.000  
4    1.000     0.000

In [55]:
# Describe the Bitcoin Sentiment
btc_df.describe()

compound   positive    neutral   negative
count  20.000000  20.000000  20.000000  20.000000
mean    0.286925   0.087700   0.893950   0.018350
std     0.426434   0.095067   0.102189   0.036977
min    -0.585900   0.000000   0.677000   0.000000
25%     0.000000   0.000000   0.821000   0.000000
50%     0.338900   0.064500   0.907500   0.000000
75%     0.642875   0.146000   1.000000   0.011750
max     0.845500   0.275000   1.000000   0.134000

In [56]:
# Describe the Ethereum Sentiment
eth_df.describe()

compound   positive    neutral   negative
count  20.000000  20.000000  20.000000  20.000000
mean   -0.024625   0.023150   0.951000   0.025850
std     0.255370   0.034562   0.053002   0.041599
min    -0.571900   0.000000   0.835000   0.000000
25%    -0.053600   0.000000   0.915000   0.000000
50%     0.000000   0.000000   0.958500   0.000000
75%     0.045025   0.050250   1.000000   0.066750
max     0.458800   0.097000   1.000000   0.111000

### Questions:

Q: Which coin had the highest mean positive score?

A: Bitcoin had the highest mean positive score with 0.0877

Q: Which coin had the highest compound score?

A: Bitcoin had the highest compound score with 0.8455

Q. Which coin had the highest positive score?

A: Bitcoin had the highest positive score with 0.2750

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [42]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [11]:
# Instantiate the lemmatizer
lemmatizer = WordNetLemmatizer()

# Create a list of stopwords
sw = set(stopwords.words('english'))

# Expand the default stopwords list if necessary
sw_addon = {'bitcoin', 'ethereum'}

# Define regex
regex = re.compile("[^a-zA-Z ]")

In [43]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
        
    # Remove the punctuation from text
    re_clean = regex.sub('', article)
   
    # Create a tokenized list of the words
    words = word_tokenize(re_clean)
    
    # Lemmatize words into root words
    lem = [lemmatizer.lemmatize(word) for word in words]
   
    # Convert the words to lowercase
    words = [word.lower() for word in words]
    
    # Remove the stop words
    tokens = [word for word in words if word not in sw]
    
    return tokens

In [13]:
# Create a new tokens column for Bitcoin
# YOUR CODE HERE!

In [14]:
# Create a new tokens column for Ethereum
# YOUR CODE HERE!

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [15]:
from collections import Counter
from nltk import ngrams

In [16]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [17]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [20]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [23]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [28]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [29]:
# Render the visualization
# YOUR CODE HERE!

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [32]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [33]:
# Render the visualization
# YOUR CODE HERE!

In [34]:
# List all Entities
# YOUR CODE HERE!

---